content/blog/2020-03-06-wasm/index.md

```diff
@@ -0,0 +1,229 @@
+
+Since WebAssembly doesn't need assumptions (such as which type certain 
+object is) during interpretting, the JavaScript engine doesn't  need to 
+bail out and reoptimize as such errors never occur. 
+
+Finally, WebAssembly also allows you to manage memory manually (it only 
+supports manual memory management as of now, but automation is to be added
+ as an option) which allows you to avoid expensive garbage collection 
+during interpretation.
+
+Therefore, WebAssembly only requires compiler backend, module loading frontend, 
+security sandboxing and supporting VM components in the existing JavaScript engine,
+and skips parsing, most compiler optimizations, bail out and garbage collection.
+
+## Implementing WebAssembly compilers
+WebAssembly is walking a tight rope between high performance (C world), and 
+safety and portability (JS world). It is hard to theorize how to optimize for
+it, so the WebAssembly team implemented extensions in different browsers to
+validate that it is indeed possible to achieve the said goal.
+
+V8 (from Chrome) and SpiderMonkey (from Firefox) reuse their optimizing JIT 
+compilers to compile ahead of time. This provides them predictable high 
+performance, as opposed to the unpredictable warmup times with JavaScript.
+Chakra (from Edge) uses lazy translation and optimizes only hot code. This 
+achieves faster startup time.
+
+To permit efficient use of baseline JIT compilers, WebAssembly is designed to
+do fast validation and ability to track registers for allocation without IR
+generation. This is done by careful design of WebAssembly instructions, which 
+the pass can use to extract register information.To integrate well with optimizing JIT compilers, WebAssembly is 
+designed to do direct-to-SSA translation (WebAssembly is not in SSA form, but offers some tools
+to derive an SSA form). Moreover, structured control flow 
+makes decoding simpler.
+
+Another key aspect in running native code in a browser is security. Sandboxing
+for C/C++ has required extensive code rewriting. However, WebAssembly is executed
+in a sandboxed environment clearly separated from the host, making it possible
+to provide some security guarantees. WebAssembly avoids performance overheads 
+through sandboxing to run at near native speed with the model created by [Native Client][nacl]
+which employs static validation of x86 machine code by mandating code generators 
+to follow certain patterns.
+
+## How good is it?
+Writing code in WebAssembly doesn't mean it'll be automatically faster. 
+JavaScript can be more performant in execution (at least for now, 
+where WebAssembly is interpreted by building into existing JavaScript engines). But 
+sometimes this requires the programmer to know JIT compilation internals and constrain to 
+one browser as each has different interpreters. WebAssembly is
+already optimized statically (more time for the compiler to optimize) by trading off 
```

Comment:
I'm still pretty confused by what "optimized statically" means. Maybe it would be good to clarify which of the two compiler mentioned above is doing the optimization.

Reply:
Yup, I clarified this description.